File: java/src/main/java/mchehab/com/java/placedetail/PlaceDetailsActivity.java
package mchehab.com.java.placedetail;

import android.os.AsyncTask;
import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

import mchehab.com.java.R;

public class PlaceDetailsActivity extends AppCompatActivity implements OnMapReadyCallback {

    private ViewPager viewPager;
    private SupportMapFragment supportMapFragment;
    private GoogleMap googleMap;
    private LinearLayout linearLayout;

    private List<Fragment> listFragments = new ArrayList<>();
    private List<String> listTitles = new ArrayList<>();

    private String placeId = "";

    private final String API_KEY = "YOUR_API_KEY_HERE";
    private final String REQUEST = "https://maps.googleapis.com/maps/api/place/details/json?key=" + API_KEY;
    private final String PHOTO_REQUEST = "https://maps.googleapis.com/maps/api/place/photo?key=" + API_KEY;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_place_details);

        linearLayout = findViewById(R.id.linearLayout);

        if (!getIntent().hasExtra("placeid")) {
            finish();
        }

        supportMapFragment = (SupportMapFragment) getSupportFragmentManager().findFragmentById(R.id.googleMap);
        supportMapFragment.getMapAsync(this);

        placeId = getIntent().getStringExtra("placeid");
        String url = REQUEST + "&placeid=" + placeId;
        new PlaceDetailsAsync().execute(url);

        listTitles.add("Reviews");
        listTitles.add("Photos");

        viewPager = findViewById(R.id.viewPager);
        TabLayout tabLayout = findViewById(R.id.tabLayout);

        tabLayout.setupWithViewPager(viewPager);
    }

    private void switchMap(int mapVisibility) {
        supportMapFragment.getView().setVisibility(mapVisibility);
        if (mapVisibility == View.VISIBLE) {
            linearLayout.setVisibility(View.GONE);
        } else {
            linearLayout.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.place_detail_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.mapListSwitcher) {
            if (item.getTitle().toString().equalsIgnoreCase("Map")) {
                switchMap(View.VISIBLE);
                item.setTitle("List");
            } else {
                switchMap(View.GONE);
                item.setTitle("Map");
            }
        } else if (item.getItemId() == android.R.id.home) {
            onBackPressed();
        }
        return true;
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        this.googleMap = googleMap;
    }

    private void setActionBarTitle(String title) {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(title);
            getSupportActionBar().setHomeButtonEnabled(true);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
    }

    private class PlaceDetailsAsync extends AsyncTask<String, Integer, JSONObject> {
        @Override
        protected JSONObject doInBackground(String... params) {
            try {
                URL url = new URL(params[0]);

                HttpURLConnection httpURLConnection = (HttpURLConnection) url.openConnection();
                httpURLConnection.setRequestMethod("GET");
                httpURLConnection.connect();

                StringBuilder stringBuilder = new StringBuilder();
                String line;
                BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(httpURLConnection.getInputStream()));
                while ((line = bufferedReader.readLine()) != null) {
                    stringBuilder.append(line);
                }

                return new JSONObject(stringBuilder.toString());
            } catch (Exception e) {
                e.printStackTrace();
            }
            return new JSONObject();
        }

        @Override
        protected void onPostExecute(JSONObject jsonObject) {
            super.onPostExecute(jsonObject);
            try {
                JSONObject result = jsonObject.getJSONObject("result");
                setActionBarTitle(result.getString("name"));
                animateGoogleMap(result);
                ArrayList<String> listReviews = getReviews(result);
                ArrayList<String> listPhotos = getPhotos(result);

                Bundle reviewsBundle = new Bundle();
                Bundle imageBundle = new Bundle();
                reviewsBundle.putStringArrayList("reviews", listReviews);
                imageBundle.putStringArrayList("photos", listPhotos);
                ImageFragment imageFragment = new ImageFragment();
                ReviewsFragment reviewsFragment = new ReviewsFragment();
                reviewsFragment.setArguments(reviewsBundle);
                imageFragment.setArguments(imageBundle);
                listFragments.add(reviewsFragment);
                listFragments.add(imageFragment);
                viewPager.setAdapter(new ViewPagerAdapter(getSupportFragmentManager(), listFragments, listTitles));
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        private void animateGoogleMap(JSONObject jsonObject) {
            try {
                String name = jsonObject.getString("name");
                JSONObject location = jsonObject.getJSONObject("geometry").getJSONObject("location");
                LatLng latLng = new LatLng(location.getDouble("lat"), location.getDouble("lng"));
                MarkerOptions markerOptions = new MarkerOptions()
                        .title(name)
                        .position(latLng);
                googleMap.animateCamera(CameraUpdateFactory.newLatLngZoom(latLng, 15.0f));
                googleMap.addMarker(markerOptions);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        private ArrayList<String> getReviews(JSONObject jsonObject) {
            ArrayList<String> reviewsList = new ArrayList<>();
            try {
                if (jsonObject.has("reviews")) {
                    JSONArray reviewsJSON = jsonObject.getJSONArray("reviews");
                    for (int i = 0; i < reviewsJSON.length(); i++) {
                        JSONObject reviewJSON = reviewsJSON.getJSONObject(i);
                        double rating = reviewJSON.getDouble("rating");
                        String text = reviewJSON.getString("text");
                        if (text.length() == 0) {
                            text = "Rating: " + rating + "/5";
                        } else {
                            text += "\nRating: " + rating + "/5";
                            ;
                        }
                        reviewsList.add(text);
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
            return reviewsList;
        }

        private ArrayList<String> getPhotos(JSONObject jsonObject) {
            ArrayList<String> list = new ArrayList<>();
            try {
                if (jsonObject.has("photos")) {
                    JSONArray jsonArray = jsonObject.getJSONArray("photos");
                    for (int i = 0; i < jsonArray.length(); i++) {
                        JSONObject json = jsonArray.getJSONObject(i);
                        int height = json.getInt("height");
                        int width = json.getInt("width");
                        String photoReference = json.getString("photo_reference");
                        String url = PHOTO_REQUEST + "&photoreference=" + photoReference + "&maxwidth=" + width + "&maxheight=" + height;
                        list.add(url);
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
            return list;
        }
    }
}

File: app/src/main/java/mchehab/com/googleplacesdetails/placedetail/PlaceDetailsActivity.kt
package mchehab.com.googleplacesdetails.placedetail

import android.os.AsyncTask
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import kotlinx.android.synthetic.main.activity_place_details.*
import mchehab.com.googleplacesdetails.R
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

class PlaceDetailsActivity : AppCompatActivity(), OnMapReadyCallback {

    private lateinit var googleMap: GoogleMap
    private lateinit var supportMapFragment: SupportMapFragment

    private val API_KEY = "YOUR_API_KEY_HERE"
    private val REQUEST = "https://maps.googleapis.com/maps/api/place/details/json?key=$API_KEY"
    private val PHOTO_REQUEST = "https://maps.googleapis.com/maps/api/place/photo?key=$API_KEY"

    private val imageFragment = ImageFragment()
    private val reviewsFragment = ReviewsFragment()
    private val listFragments = listOf(reviewsFragment, imageFragment)
    private val listTitles = listOf("Reviews", "Images")

    private var placeId = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_place_details)

        tabLayout.setupWithViewPager(viewPager)

        //this should never be true
        if (intent.hasExtra("placeId")){
            placeId = intent.getStringExtra("placeId")

        }else{
            //show alert dialog and finish activity
            AlertDialog.Builder(this)
                .setTitle("Error")
                .setMessage("Cannot retrieve place, please try again later")
                .setPositiveButton("Ok") { dialog, which ->
                    dialog.dismiss()
                    finish()
                }
                .create()
                .show()
        }

        supportMapFragment = supportFragmentManager.findFragmentById(R.id.googleMap) as SupportMapFragment
        supportMapFragment.getMapAsync(this)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.place_detail_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (item?.itemId == R.id.mapListSwitcher){
            if (item.title.toString().equals("List", true)){
                switchMap(View.GONE)
                item.title = "Map"
            }else{
                switchMap(View.VISIBLE)
                item.title = "List"
            }
        }else if (item?.itemId == android.R.id.home){
            onBackPressed()
        }
        return true
    }

    override fun onMapReady(googleMap: GoogleMap) {
        this.googleMap = googleMap
        val url = "$REQUEST&placeid=$placeId"
        PlaceDetailsAsync().execute(url)
    }

    private fun switchMap(mapVisibility: Int){
        supportMapFragment.view?.visibility = mapVisibility
        if (mapVisibility == View.VISIBLE){
            linearLayout.visibility = View.GONE
        }else{
            linearLayout.visibility = View.VISIBLE
        }
    }

    private inner class PlaceDetailsAsync : AsyncTask<String, Int, JSONObject>() {
        override fun doInBackground(vararg params: String): JSONObject {
            val url = URL(params[0])
            val httpURLConnection = url.openConnection() as HttpURLConnection
            httpURLConnection.requestMethod = "GET"
            httpURLConnection.connect()
            val placeDetails = httpURLConnection.inputStream.bufferedReader().readText()
            return JSONObject(placeDetails)
        }

        override fun onPostExecute(jsonObject: JSONObject) {
            try {
                super.onPostExecute(jsonObject)
                val result = jsonObject.getJSONObject("result")
                setupActionBar(result.getString("name"))
                animateGoogleMap(result)
                val reviewsList = getReviews(result)
                val urls = getPhotos(result)
                val reviewsBundle = Bundle()
                val imageBundle = Bundle()
                reviewsBundle.putStringArrayList("reviews", reviewsList)
                imageBundle.putStringArrayList("photos", urls)
                reviewsFragment.arguments = reviewsBundle
                imageFragment.arguments = imageBundle
                viewPager.adapter = ViewPagerAdapter(
                    supportFragmentManager,
                    listFragments,
                    listTitles
                )
            } catch (exception: Exception) {
                exception.printStackTrace()
            }
        }

        private fun setupActionBar(title: String){
            supportActionBar?.title = title
            supportActionBar?.setHomeButtonEnabled(true)
            supportActionBar?.setDisplayHomeAsUpEnabled(true)
        }

        private fun animateGoogleMap(jsonObject: JSONObject){
            val name = jsonObject.getString("name")
            val location = jsonObject.getJSONObject("geometry").getJSONObject("location")
            val latLng = LatLng(location.getDouble("lat"), location.getDouble("lng"))
            googleMap.animateCamera(CameraUpdateFactory.newLatLngZoom(latLng, 15.0f))
            val marker = MarkerOptions()
                    .title(name)
                    .position(latLng)
            googleMap.addMarker(marker)
        }

        private fun getReviews(jsonObject: JSONObject): ArrayList<String>{
            val reviewsList = ArrayList<String>()
            if(jsonObject.has("reviews")){
                val reviews = jsonObject.getJSONArray("reviews")
                for (i in 0 until reviews.length()){
                    val reviewJSON = reviews.getJSONObject(i)
                    var text = reviewJSON.getString("text")
                    val rating = reviewJSON.getDouble("rating")
                    if (text.isEmpty()){
                        text = "Rating: $rating/5"
                    }else{
                        text = "$text\nRating: $rating/5"
                    }
                    reviewsList.add(text)
                }
            }

            return reviewsList
        }

        private fun getPhotos(jsonObject: JSONObject): ArrayList<String>{
            val urls = ArrayList<String>()
            if (jsonObject.has("photos")){
                val photos = jsonObject.getJSONArray("photos")
                for (i in 0 until photos.length()) {
                    val json = photos.getJSONObject(i)
                    val height = json.getInt("height")
                    val width = json.getInt("width")
                    val photoReference = json.get("photo_reference")
                    val url = "$PHOTO_REQUEST&photoreference=$photoReference&maxwidth=$width&maxheight=$height"
                    urls.add(url)
                }
            }
            return urls
        }
    }
}